muxink/src/demux.rs

```diff
@@ -0,0 +1,325 @@
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::{stream::Fuse, Stream, StreamExt};
+
+/// A frame demultiplexer.
+///
+/// A demultiplexer is not used directly, but used to spawn demultiplexing handles.
+pub struct Demultiplexer<S> {
+    /// The underlying `Stream`, `Fuse`d in order to make it safe to be called once its output
+    /// `Poll::Ready(None)`.
+    stream: Fuse<S>,
+    /// Holds the frame and channel, if available, which has been read by a `DemultiplexerHandle`
+    /// corresponding to a different channel.
+    next_frame: Option<(u8, Bytes)>,
+    /// A bit-field representing the channels which have had `DemultiplexerHandle`s constructed.
+    active_channels: [u8; 32],
+}
+
+impl<S: Stream> Demultiplexer<S> {
+    /// Creates a new demultiplexer with the given underlying stream.
+    pub fn new(stream: S) -> Demultiplexer<S> {
+        Demultiplexer {
+            // We fuse the stream in case its unsafe to call it after yielding `Poll::Ready(None)`
+            stream: stream.fuse(),
+            // Initially, we have no next frame
+            next_frame: None,
+            // Initially, all channels are inactive
+            active_channels: [0b00000000; 32],
+        }
+    }
+}
+
+// Here, we write the logic for accessing and modifying the bit-field representing the active
+// channels.
+impl<S> Demultiplexer<S> {
+    fn activate_channel(&mut self, channel: u8) {
+        self.active_channels[(channel / 8) as usize] |=
+            2u8.checked_pow((channel % 8) as u32).unwrap();
+    }
+
+    fn deactivate_channel(&mut self, channel: u8) {
+        // TODO Single operation instead of two.
+        if self.channel_is_active(channel) {
+            self.active_channels[(channel / 8) as usize] ^=
+                2u8.checked_pow((channel % 8) as u32).unwrap();
+        }
+    }
+
+    fn channel_is_active(&self, channel: u8) -> bool {
+        (self.active_channels[(channel / 8) as usize]
+            & 2u8.checked_pow((channel % 8) as u32).unwrap())
+        .count_ones()
+            == 1
+    }
+}
+
+impl<S: Stream> Demultiplexer<S> {
+    /// Creates a handle listening for frames on the given channel.
+    ///
+    /// Any item on this channel sent to the `Stream` underlying the `Demultiplexer` we used to
+    /// create this handle will be read only when all other messages for other channels have been
+    /// read first. If one has handles on the same channel created via the same underlying
+    /// `Demultiplexer`, each message on that channel will only be received by one of the handles.
+    /// Unless this is desired behavior, this should be avoided.
+    pub fn create_handle(demux: Arc<Mutex<Self>>, channel: u8) -> DemultiplexerHandle<S> {
+        let mut guard = match demux.as_ref().try_lock() {
+            Err(_err) => panic!("TODO"),
+            Ok(guard) => guard,
+        };
+
+        if guard.channel_is_active(channel) {
+            panic!("TODO")
```

Comment:
Not relevant to this PR, but for future efforts, you can use `todo!("need to implement X and Y")`.